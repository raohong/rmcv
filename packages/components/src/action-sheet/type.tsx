export type ActionSheetAction = {
  /**
   * @description 标题
   */
  name: React.ReactNode;
  /**
   * @description 二级标题
   */
  subName?: React.ReactNode;
  /**
   * @description 文字颜色
   */
  color?: string;
  /**
   * @description 选项自定义 class
   */
  className?: string;
  /**
   * @description 是否为加载状态
   */
  loading?: boolean;
  /**
   * @description 是否为禁用状态
   */
  disabled?: boolean;
  /**
   * @description 点击时触发的回调函数
   */
  callback?: (action: ActionSheetAction) => void;
};

export type ActionSheetProps = {
  /**
   * @description 是否显示动作面板
   */
  visible?: boolean;
  /**
   * @description 面板选项列表
   */
  actions?: ActionSheetAction[];
  /**
   * @description 动作面板标题
   */
  title?: React.ReactNode;
  /**
   * @description 动作面板描述
   */
  description?: React.ReactNode;
  /**
   * @description 关闭按钮文字
   */
  cancelText?: React.ReactNode;
  /**
   * @description 是否可关闭的
   * @default true
   */
  closable?: boolean;
  /**
   * @description 关闭图标
   */
  closeIcon?: React.ReactNode;
  /**
   * @description 是否显示圆角
   * @default true
   */
  round?: boolean;
  /**
   * @description 自定义 class
   */
  className?: string;
  /**
   * @description 是否显示 overlay
   * @default true
   */
  overlay?: boolean;
  /**
   * @description overlay 自定义 className
   */
  overlayClassName?: string;
  /**
   * @description overlay 自定义 style
   */
  overlayStyle?: React.CSSProperties;
  /**
   * @description overlay 点击是否关闭
   * @default true
   */
  overlayClosable?: boolean;
  /**
   * @description 是否在显示弹层时才渲染节点
   * @default true
   */
  lazyRender?: boolean;
  /**
   * @description
   * @default true
   */
  lockScroll?: boolean;
  /**
   * @description 是否在页面回退时自动关闭
   */
  closeOnPopState?: boolean;
  /**
   * @description 是否开启底部安全区适配
   */
  safeArea?: boolean;
  /**
   * @description 点击选项后是否关闭
   */
  closeOnClickAction?: boolean;
  /**
   * @description 点击选项时触发，禁用或加载状态下不会触发
   */
  onSelect?: (action: ActionSheetAction, index: number) => void;
  /**
   * @description 关闭回调
   */
  onClose?: () => void;
  /**
   * @description 点击取消按钮，此时还要触发 onClose
   */
  onCancel?: () => void;
  /**
   * @description 点击 overlay 回调
   */
  onOverlayClick?: () => void;
  /**
   * @description 关闭前的回调函数，返回 false 可阻止关闭，支持返回 Promise
   */
  onBeforeClose?: (
    action: ActionSheetAction,
  ) => void | boolean | Promise<boolean>;
  /**
   * @description 动画结束后回调
   */
  afterVisibleChange?: (visible: boolean) => void;
  /**
   * @description 自定义内容
   */
  children?: React.ReactNode;
};
