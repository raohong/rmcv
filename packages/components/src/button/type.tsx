import type { JSXIntrinsicElementProps } from '../types';
import type { LoadingProps } from '../loading';

export type ButtonSize = 'large' | 'default' | 'small' | 'mini';

export type ButtonType = 'primary' | 'success' | 'warning' | 'danger' | 'link';

export type ButtonShape = 'round' | 'square';

export type NativeButtonHTMLType = 'button' | 'submit' | 'reset';

type BaseButtonProps = {
  /**
   * @description 按钮类型
   */
  type?: ButtonType;
  /**
   * @description 按钮尺寸
   */
  size?: ButtonSize;
  /**
   * @description 是否是块级元素
   */
  block?: boolean;
  /**
   * @description 是否镂空
   */
  plain?: boolean;
  /**
   * @description 按钮形状
   */
  shape?: ButtonShape;
  /**
   * @description 按钮 disabled
   */
  disabled?: boolean;
  /**
   * @description 按钮 icon
   */
  icon?: React.ReactElement;
  /**
   * @description 是否 loading
   */
  loading?: boolean;
  /**
   * @description border 是否是高清
   */
  hairline?: boolean;
  /**
   * @@description 点击事件
   */
  onClick?: React.MouseEventHandler<HTMLElement>;
  /**
   * @@description 按钮内容
   */
  children?: React.ReactNode;
  /**
   * @@description loading 图标文字
   */
  loadingText?: string;
  /**
   * @@description loading 图标类型
   */
  loadingType?: LoadingProps['type'];
  /**
   * @@description loading 图标大小
   * @default 20px
   */
  loadingSize?: LoadingProps['size'];
};

type AnchorButtonProps = {
  /**
   * @description 按钮 html 链接
   */
  href?: string;
  target?: string;
} & BaseButtonProps &
  JSXIntrinsicElementProps<'button', 'onClick' | 'type'>;

type NativeButtonProps = {
  /**
   * 按钮的原始 type
   */
  htmlType?: NativeButtonHTMLType;
} & BaseButtonProps &
  JSXIntrinsicElementProps<'button', 'onClick' | 'type'>;

export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>;
