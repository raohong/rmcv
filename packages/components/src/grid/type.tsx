import type { IntrinsicElementsKeys, JSXIntrinsicElementProps } from '../types';

export type GridDirection = 'vertical' | 'horizontal';

export type GridProps = {
  /**
   * @description 列数
   * @default 3
   */
  column?: number;
  /**
   * @description 图标大小
   * @default 28
   */
  iconSize?: number | string;
  /**
   * @description 格子之间的间距
   */
  gutter?: number | string;
  /**
   * @description 是否显示边框
   */
  border?: boolean;
  /**
   * @description 是否将格子内容居中显示
   */
  center?: boolean;
  /**
   * @description 是否将格子固定为正方形
   */
  square?: boolean;
  /**
   * @description 是否开启格子点击反馈
   */
  clickable?: boolean;
  /**
   * @description 格子内容排列的方向
   * @default vertical
   */
  direction?: GridDirection;
  /**
   * @description 渲染组件
   * @default div
   */
  component?: IntrinsicElementsKeys;
} & JSXIntrinsicElementProps<'div'>;

export type GridItemProps = {
  /**
   * @description 文字内容
   */
  text?: React.ReactNode;
  /**
   * @description 图标
   */
  icon?: React.ReactNode;
  /**
   * @description 图标大小
   * @default 28
   */
  iconSize?: number | string;
  /**
   * @description 是否显示徽标小红点
   */
  dot?: boolean;
  /**
   * @description 右上角徽标的内容
   */
  badge?: number | string;
  /**
   * @description 徽标内容为数字时，最大值
   * @default 99
   */
  max?: number;
  /**
   * @description 当 content 是 0 时，是否显示
   * @default true
   */
  showZero?: boolean;
  /**
   * @description 渲染组件
   * @default div
   */
  component?: IntrinsicElementsKeys;
  /**
   * @description content 自定义
   */
  contentClassName?: string;
} & JSXIntrinsicElementProps<'div'>;
