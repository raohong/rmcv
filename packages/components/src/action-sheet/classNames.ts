import { generateComponentClassNameUtility } from '@rmc-vant/system';
import {
  ActionSheetComponentState,
  ActionSheetNSlot,
  ActionSheetSlot,
} from './interface';

export const ActionSheetName = 'ActionSheet';

export const {
  actionSheetClassNames,
  composeActionSheetSlotClassNames,
  getActionSheetSlotClassNames,
  getActionSheetClassName,
} = generateComponentClassNameUtility<
  typeof ActionSheetName,
  ActionSheetComponentState,
  ActionSheetSlot,
  ActionSheetNSlot
>(
  ActionSheetName,
  {
    root: true,
    title: true,
    description: true,
    content: true,
    cancelButton: true,
    item: true,
    itemSubname: true,
    loadingIcon: true,
    closeIcon: true,
    itemDisabled: true,
    itemLoading: true,
  },
  () => ({
    root: ['root'],
    title: ['title'],
    closeIcon: ['closeIcon'],
    description: ['description'],
    content: ['content'],
    cancelButton: ['cancelButton'],
    item: ['item'],
    itemSubname: ['itemSubname'],
    loadingIcon: ['loadingIcon'],
  }),
);
