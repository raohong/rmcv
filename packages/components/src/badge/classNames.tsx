import { generateComponentClassNameUtility } from '@rmc-vant/system';
import { camelCase } from '@rmc-vant/utils';
import type { BadgeComponentState, BadgeNSlot, BadgeSlot } from './interface';

export const BadgeName = 'Badge';

export const {
  badgeClassNames,
  getBadgeSlotClassNames,
  composeBadgeSlotClassNames,
} = generateComponentClassNameUtility<
  typeof BadgeName,
  BadgeComponentState,
  BadgeSlot,
  BadgeNSlot
>(
  BadgeName,
  {
    root: true,
    number: true,
    numberWrapper: true,
    wrapper: true,
    dot: true,
    positionBottomLeft: true,
    positionBottomRight: true,
    positionTopLeft: true,
    positionTopRight: true,
  },
  ({ dot, position }) => ({
    root: ['root', camelCase(`position-${position}`)],
    wrapper: ['wrapper', dot && 'dot'],
    number: ['number'],
    numberWrapper: ['numberWrapper'],
  }),
);
