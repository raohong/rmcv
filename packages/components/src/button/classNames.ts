import { generateComponentClassNameUtility } from '@rmc-vant/system';
import { camelCase } from '@rmc-vant/utils';
import type { ButtonComponentState, ButtonNSlot, ButtonSlot } from './interface';

export const ButtonName = 'Button';

export const {
  buttonClassNames,
  getButtonSlotClassNames,
  composeButtonSlotClassNames,
} = generateComponentClassNameUtility<
  typeof ButtonName,
  ButtonComponentState,
  ButtonSlot,
  ButtonNSlot
>(
  ButtonName,
  {
    root: true,
    icon: true,
    loading: true,
    disabled: true,
    outlined: true,
    outlinedSizeMini: true,
    outlinedSizeSmall: true,
    outlinedSizeLarge: true,
    outlinedPrimary: true,
    outlinedSuccess: true,
    outlinedWarning: true,
    outlinedDanger: true,
    outlinedShapeRound: true,
    outlinedShapeSquare: true,
    contained: true,
    containedSizeMini: true,
    containedSizeSmall: true,
    containedSizeLarge: true,
    containedPrimary: true,
    containedSuccess: true,
    containedWarning: true,
    containedDanger: true,
    containedShapeRound: true,
    containedShapeSquare: true,
    loadingIcon: true,
    block: true,
  },
  ({ variant, size, shape, loading, disabled, type, block }) => ({
    root: [
      'root',
      variant,
      loading && 'loading',
      disabled && 'disabled',
      size !== 'normal' && camelCase(`${variant}-size-${size}`),
      type !== 'default' && camelCase(`${variant}-${type}`),
      shape !== 'default' && camelCase(`${variant}-shape-${shape}`),
      block && 'block',
    ],
    icon: ['icon'],
    loadingIcon: ['loadingIcon'],
  }),
);
