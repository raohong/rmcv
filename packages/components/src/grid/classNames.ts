import { generateComponentClassNameUtility } from '@rmc-vant/system';
import type { GridComponentState, GridNSlot, GridSlot } from './interface';

export const GridName = 'Grid';

export const { gridClassNames, composeGridSlotClassNames, getGridSlotClassNames }
  = generateComponentClassNameUtility<
    typeof GridName,
    GridComponentState,
    GridSlot,
    GridNSlot
  >(
    GridName,
    {
      root: true,
      item: true,
      horizontal: true,
      vertical: true,
      square: true,
      itemContent: true,
      itemIcon: true,
      itemLabel: true,
    },
    ({ square, direction }) => ({
      root: ['root', square && 'square', direction],
      item: ['item'],
      itemContent: ['itemContent'],
      itemLabel: ['itemLabel'],
      itemIcon: ['itemIcon'],
    }),
  );
