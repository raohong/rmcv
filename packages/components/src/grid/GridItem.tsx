import { isEmpty, isNil } from '@rmc-vant/utils';
import clsx from 'clsx';
import React from 'react';
import { Badge } from '../badge';
import type { GridItemProps } from './interface';
import { GridIcon, GridLabel, StyledGridItem, StyledGridItemContent } from './styles';

const GridItem = React.forwardRef<HTMLDivElement, GridItemProps>(
  (
    {
      label,
      icon,
      dot,
      max,
      showZero,
      className,
      badge,
      contentClassName,
      componentState,
      slotClassNames,
      component = 'div',
      children,
      ...rest
    },
    ref,
  ) => {
    const content = (
      <StyledGridItemContent
        componentState={componentState}
        className={clsx(slotClassNames.itemContent, contentClassName)}
        activeStyle={({ theme }) => ({
          '&[class]': {
            background: theme.palette.active,
          },
        })}
        // @ts-ignore
        component='div'
        disabled={!componentState.clickable}
        {...(componentState.clickable && { role: 'button', tabIndex: 0 })}
      >
        {isNil(children)
          ? (
              <>
                <Badge max={max} dot={dot} content={badge} showZero={showZero}>
                  <GridIcon
                    componentState={componentState}
                    className={slotClassNames.itemIcon}
                  >
                    {icon}
                  </GridIcon>
                </Badge>
                {!isEmpty(label) && (
                  <GridLabel
                    className={slotClassNames.itemLabel}
                    componentState={componentState}
                  >
                    {label}
                  </GridLabel>
                )}
              </>
            )
          : (
              children
            )}
      </StyledGridItemContent>
    );

    return (
      <StyledGridItem
        ref={ref}
        as={component}
        className={clsx(slotClassNames.item, className)}
        componentState={componentState}
        {...rest}
      >
        {content}
      </StyledGridItem>
    );
  },
);

export default GridItem;
