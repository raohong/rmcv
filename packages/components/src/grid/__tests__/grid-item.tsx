import { render, screen } from '@testing-library/react';
import GridItem from '../GridItem';
import { gridClassNames } from '../classNames';
import type { GridComponentState } from '../interface';

const testId = 'grid-item';

const state = {
  clickable: false,
  gutter: 0,
  column: 1,

  direction: 'horizontal',
  center: false,
  square: false,
  border: false,
  iconSize: 20,
} satisfies GridComponentState;

const states = {
  slotClassNames: gridClassNames,
  componentState: state,
};

it('render correctly', () => {
  const tree = render(<GridItem {...states} />);

  expect(tree.asFragment()).toMatchSnapshot();
});

it('render with icon', () => {
  render(<GridItem data-testid={testId} icon='text-icon' {...states} />);

  expect(screen.getByText('text-icon')).toBeInTheDocument();
});

it('render with iconSize', () => {
  render(
    <GridItem
      data-testid={testId}
      icon='text-icon'
      {...states}
      componentState={{
        ...states.componentState,
        iconSize: 24,
      }}
    />,
  );

  expect(screen.getByText('text-icon')).toHaveStyleRule('font-size', '24px');
});

it('render with label', () => {
  render(<GridItem data-testid={testId} label='text' {...states} />);

  expect(screen.getByText('text')).toHaveClass(gridClassNames.itemText);
});

it('render with contentClassName', () => {
  render(<GridItem data-testid={testId} contentClassName='test' {...states} />);

  expect(screen.getByTestId(testId).querySelector('.test')).not.toBeNull();
});
