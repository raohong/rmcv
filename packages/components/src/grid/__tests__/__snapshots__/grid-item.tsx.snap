// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  box-sizing: border-box;
  max-width: 100%;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: inherit;
  border: 0;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
  cursor: not-allowed;
  position: relative;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  padding: 12px 16px;
  background: #fff;
  gap: 8px;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  font-family: -apple-system-font,Helvetica Neue,Arial,sans-serif;
  line-height: 14px;
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}

.emotion-3 {
  font-size: 20px;
  width: 1em;
  height: 1em;
  box-sizing: border-box;
}

<div
    class="RVuiGrid-item emotion-0"
  >
    <div
      class="RVuiTouchable-root RVuiGrid-itemContent emotion-1"
      role="button"
      type="button"
    >
      <div
        class="RVuiBadge-root RVuiBadge-positionTopRight emotion-2"
      >
        <div
          class="RVuiGrid-itemIcon emotion-3"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
